### PR TITLE
API - 3495  Investigate Webpack Config and Plugins

### PR DESCRIPTION
**Pull request by TS000**
### Description
API - 3495 Jira ticket [here](https://vajira.max.gov/browse/API-3495)

Update Webpack plugins that are either out of date or no longer useful so that we can have the most efficient and maintainable Webpack config going forward.

Tickets for plugins that need to be updated:
* [API - 4493: Webpack](https://vajira.max.gov/browse/API-4493) 
* [API - 4498: React-Dev-Utils](https://vajira.max.gov/browse/API-4498) 
* [API - 4495: CopyPlugin](https://vajira.max.gov/browse/API-4495)

### Process

Worked through the plugins used in the 3 webpack.config files we currently use.

# Webpack Plugin Audit
### This is a check to see which plugins are being used, need updated, or are severely exploitable.

Plugins moved to latest release were done so only if there were no issues with the upgrade. Therefore the plugins that are not up to date are due to various issues with the latest release and other parts of the Dev Portal.

## webpack.config.dev.js

Plugin             | Link  | Use it?  | To date?| Security bad?   | 
-----------------|---------|----------|---------|---------|
autoprefixer     |[NPM](https://www.npmjs.com/package/autoprefixer) |    Yes |   No |No |                   
path     |[NPM](https://www.npmjs.com/package/path) |    Yes |   Yes |   No |     
webpack     |[NPM](https://www.npmjs.com/package/webpack) |    Yes |   No |   No |
CaseSensitivePathsPlugin     |[NPM](https://www.npmjs.com/package/case-sensitive-paths-webpack-plugin) |    Yes |   Yes |   [Yes](https://snyk.io/advisor/npm-package/case-sensitive-paths-webpack-plugin) |
react-dev-utils/InterpolateHtmlPlugin     |[NPM](https://www.npmjs.com/package/react-dev-utils) |    Yes |   No |   [Yes](https://snyk.io/test/npm/react-dev-utils/6.1.1) |
react-dev-utils/WatchMissingNodeModulesPlugin     |[NPM](https://www.npmjs.com/package/react-dev-utils) |    Yes |   No |   [Yes](https://snyk.io/test/npm/react-dev-utils/6.1.1) |
react-dev-utils/ModuleScopePlugin     |[NPM](https://www.npmjs.com/package/react-dev-utils) |    Yes |   No |   [Yes](https://snyk.io/test/npm/react-dev-utils/6.1.1) |
ForkTsCheckerWebpackPlugin     |[NPM](https://www.npmjs.com/package/fork-ts-checker-webpack-plugin) |    Yes |   Yes |   No |
TsconfigPathsPlugin     |[NPM](https://www.npmjs.com/package/tsconfig-paths-webpack-plugin) |    Yes |   Yes | No

--------------------------
## webpack.config.prod.js
Plugin             | Link  | Use it?  | To date?| Security bad?   | 
-----------------|---------|----------|---------|---------|
react-dev-utils/ModuleScopePlugin     |[NPM](https://www.npmjs.com/package/react-dev-utils) |    Yes |   No |   [Yes](https://snyk.io/test/npm/react-dev-utils/6.1.1) 
HtmlWebpackPlugin     |[NPM](https://www.npmjs.com/package/html-webpack-plugin) |   Yes |   Yes | No
react-dev-utils/InlineChunkHtmlPlugin     |[NPM](https://www.npmjs.com/package/react-dev-utils) |    Yes |   No |   [Yes](https://snyk.io/test/npm/react-dev-utils/6.1.1)
MiniCssExtractPlugin     |[NPM](https://www.npmjs.com/package/mini-css-extract-plugin) |    Yes |   Yes | No
OptimizeCSSAssetsPlugin     |[NPM](https://www.npmjs.com/package/optimize-css-assets-webpack-plugin) |    Yes |   Yes | [No](https://snyk.io/advisor/npm-package/optimize-css-assets-webpack-plugin)
ManifestPlugin     |[NPM](https://www.npmjs.com/package/webpack-manifest-plugin) |    Yes |   No | [No](https://snyk.io/advisor/npm-package/webpack-manifest-plugin)
react-dev-utils/InterpolateHtmlPlugin     |[NPM](https://www.npmjs.com/package/react-dev-utils) |    Yes |   No |   [Yes](https://snyk.io/test/npm/react-dev-utils/6.1.1) |
WorkboxWebpackPlugin     |[NPM](https://www.npmjs.com/package/workbox-webpack-plugin) |   Yes |   Yes | [No](https://snyk.io/advisor/npm-package/workbox-webpack-plugin)
getCSSModuleLocalIdent     |[NPM](https://www.npmjs.com/package/react-dev-utils) |    Yes |   No |   [Yes](https://snyk.io/test/npm/react-dev-utils/6.1.1) |
ForkTsCheckerWebpackPlugin     |[NPM](https://www.npmjs.com/package/fork-ts-checker-webpack-plugin) |    Yes |   Yes |   No |
react-dev-utils/InlineChunkHtmlPlugin     |[NPM](https://www.npmjs.com/package/react-dev-utils) |    Yes |   No |   [Yes](https://snyk.io/test/npm/react-dev-utils/6.1.1) |
CopyPlugin     |[NPM](https://www.npmjs.com/package/copy-webpack-plugin) |    Yes |   No |   [No](https://snyk.io/vuln/npm:copy-webpack-plugin) |
TsconfigPathsPlugin     |[NPM](https://www.npmjs.com/package/tsconfig-paths-webpack-plugin) |    Yes |   Yes | No|
-----
## webpackDevServer.config.js
Plugin             | Link  | Use it?  | To date?| Security bad?   | 
-----------------|---------|----------|---------|---------|
react-dev-utils/errorOverlayMiddleware     |[NPM](https://www.npmjs.com/package/react-dev-utils) |    Yes |   No |   [Yes](https://snyk.io/test/npm/react-dev-utils/6.1.1)|
react-dev-utils/noopServiceWorkerMiddleware |[NPM](https://www.npmjs.com/package/react-dev-utils) |    Yes |   No |   [Yes](https://snyk.io/test/npm/react-dev-utils/6.1.1)|
react-dev-utils/ignoredFiles |[NPM](https://www.npmjs.com/package/react-dev-utils) |    Yes |   No |   [Yes](https://snyk.io/test/npm/react-dev-utils/6.1.1)|

### Requested feedback

All feedback welcome.

### CODE REVIEW

File: config/webpack.config.dev.js [hunk 1]
```diff
@@ -106,6 +106,7 @@ module.exports = {
       // please link the files into your node_modules/ and let module-resolution kick in.
       // Make sure your source files are compiled, as they will not be processed in any way.
       new ModuleScopePlugin(paths.appSrc, [paths.appPackageJson]),
+      // This module helps us get past the need alias entries in the webpack.config file we don't use.
```
Comment (anthony-kopczyk) [inline config/webpack.config.dev.js:109]: I think there's a typo in this sentence - an extra word maybe?

**Review by anthony-kopczyk — COMMENTED** [review]
This looks good to me, but I will for others to approve as well.

File: config/webpack.config.dev.js (hunk 1 above)
Comment (TS000) [reply config/webpack.config.dev.js:109]: It's just convoluted. I'll clean it up ⭐

File: package.json [hunk 2]
```diff
@@ -114,9 +114,9 @@
     "babel-loader": "8.0.4",
     "babel-plugin-named-asset-import": "^0.2.3",
     "babel-preset-react-app": "^7.0.2",
-    "case-sensitive-paths-webpack-plugin": "2.1.1",
+    "case-sensitive-paths-webpack-plugin": "^2.3.0",
     "chalk": "1.1.3",
-    "copy-webpack-plugin": "^6.2.1",
+    "copy-webpack-plugin": "^6.4.1",
```
Comment (mjlumetta) [inline package.json:119]: guessing there's a reason for not upgrading to 7.0.0? saw you made the ticket, but the table's out of sync with that, i think. might be good to link [the medium severity indirect vulnerability](https://snyk.io/test/npm/copy-webpack-plugin/6.4.1) from there if others are going to be reviewing.

**Review by mjlumetta — COMMENTED** [review]
really nice job on those tables!! 💅 think this is looking really good, thanks for combing through all of that. gonna throw out a few questions that don't go anywhere in the code (some of these are about updating the PR description, in which case feel free to ignore if we're close to closing this out/not necessary if there aren't going to be new reviewers).

1. `resolve` - can we uninstall this package as a direct dependency? seems like we're using Node's native `require.resolve()` rather than the module. if that's the case, we can cancel API-4496 as well.
2. worth updating the `ManifestPlugin` line in the table to reflect rolling back the changes?
3. `getCSSModuleLocalIdent` looks like it's from `react-dev-utils`. if that's the case, can API-4500 be canceled?

File: package.json (hunk 2 above)
Comment (TS000) [reply package.json:119]: Good eye!

File: package.json [hunk 3]
```diff
@@ -114,9 +114,9 @@
     "babel-loader": "8.0.4",
     "babel-plugin-named-asset-import": "^0.2.3",
     "babel-preset-react-app": "^7.0.2",
-    "case-sensitive-paths-webpack-plugin": "2.1.1",
+    "case-sensitive-paths-webpack-plugin": "^2.3.0",
```
[the remark this answers is not in the document]
Comment (TS000) [reply package.json:117]: Fixed, thank you!

**Comment by TS000** [comment]
> really nice job on those tables!! 💅 think this is looking really good, thanks for combing through all of that. gonna throw out a few questions that don't go anywhere in the code (some of these are about updating the PR description, in which case feel free to ignore if we're close to closing this out/not necessary if there aren't going to be new reviewers).
> 
> 1. `resolve` - can we uninstall this package as a direct dependency? seems like we're using Node's native `require.resolve()` rather than the module. if that's the case, we can cancel API-4496 as well.
> 2. worth updating the `ManifestPlugin` line in the table to reflect rolling back the changes?
> 3. `getCSSModuleLocalIdent` looks like it's from `react-dev-utils`. if that's the case, can API-4500 be canceled?
> 4. something is out of sync between `package-lock.json` and `package.json` for `workbox-webpack-plugin`; `package.json` is still referencing the old version, and `package-lock.json` is referencing the most recent version (which is what the table documents). that inconsistency appears to be causing the CI failure:
> 
> ```
> [2021-01-19T17:23:48.411Z] + npm audit --audit-level moderate
> [2021-01-19T17:23:48.666Z] npm ERR! code ELOCKVERIFY
> [2021-01-19T17:23:48.666Z] npm ERR! Errors were found in your package-lock.json, run  npm install  to fix them.
> [2021-01-19T17:23:48.666Z] npm ERR!     Invalid: lock file's workbox-webpack-plugin@6.0.2 does not satisfy workbox-webpack-plugin@3.6.3
> ```

I've addressed the issues! Thanks for the help ⭐

**Pull request closed without merging**